coremltools/converters/nnssa/coreml/ssa_converter.py

```diff
@@ -731,6 +732,126 @@ def _convert_custom_layer(self, node):
             raise ValueError('Custom Shape Function for {} not provided!'.format(node.op))
         shapes.propagate_single_layer(layer, self.tensor_shapes, custom_shape_function=self.custom_shape_functions[node.op])
 
+    def _convert_einsum(self, node):
+
+        input_nodes, input_names, input_types = self._get_input_tensors(node)
+        if len(input_names) > 2:
+            raise ValueError("currently, 'einsum' operation is only supported when it has less than equal to 2 inputs.")
+        equation = node.attr.get('equation')
+        if not '->' in equation:
+            raise ValueError('current einsum does not support matrix diagonal operations.')
+
+        # Helper function
+        def get_transpose_map(prefix, suffix):
+            prefix_map = dict(zip(prefix,range(len(prefix))))
+            return [prefix_map[x] for x in suffix]
+
+        # Parse equation
+        prefix = equation.split('->')[0]
+        suffix = equation.split('->')[1]
+
+        # Pattern matching
+        builder = self._get_builder()
+        if not ',' in prefix:
+            # Transpose
+            axes = get_transpose_map(prefix, suffix)
+            layer = builder.add_transpose(
+                    name = node.name,
+                    axes = axes,
+                    input_name = input_names[0],
+                    output_name = node.name)
+            shapes.propagate_single_layer(layer, self.tensor_shapes)
+
+        else:
+            a, b = prefix.split(',')
+            if not len(a) == len(b):
+                raise ValueError("currently 'einsum' operation is only supported when both inputs have the same rank.")
+            if suffix == '':
+                # Inner Product
```

Comment:
FYI, we have an `InnerProduct` layer. Seems you have it divided into multiple layers, is that necessary?